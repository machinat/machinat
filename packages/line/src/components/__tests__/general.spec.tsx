/* eslint-disable no-await-in-loop */
import Sociably from '@sociably/core';
import Renderer from '@sociably/core/renderer';
import generalComponentDelegator from '../general.js';

const renderer = new Renderer('line', generalComponentDelegator);

describe('<p/>', () => {
  test('<p/> renders into individual text segment', async () => {
    const segments = await renderer.render(
      <>
        <p>foo</p>
        bar
        <p>baz</p>
      </>,
      null,
      null
    );

    expect(segments).toMatchSnapshot();
    expect(segments?.map((seg) => seg.value)).toMatchInlineSnapshot(`
      [
        "foo",
        "bar",
        "baz",
      ]
    `);
  });

  it('return null if content is empty', async () => {
    await expect(renderer.render(<></>, null, null)).resolves.toBe(null);
  });

  it('throw if non-text children received', async () => {
    expect(
      renderer.render(
        <p>
          <img />
        </p>,
        null,
        null
      )
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <p/>"`
    );
  });
});

describe('text components', () => {
  test('render shallow elements', async () => {
    const results = await Promise.all(
      [
        <b>important</b>,
        <i>italic</i>,
        <s>nooooo</s>,
        <u>underlined</u>,
        <code>foo.bar()</code>,
        <pre>foo.bar('hello world')</pre>,
        <br />,
      ].map((element) => renderer.render(element, null, null))
    );

    expect(results).toMatchSnapshot();
    expect(results.flat().map((seg) => seg?.value)).toMatchInlineSnapshot(`
      [
        "important",
        "italic",
        "nooooo",
        "underlined",
        "foo.bar()",
        "foo.bar('hello world')",
        "
      ",
      ]
    `);
  });

  test('render nested elements', async () => {
    const promise = renderer.render(
      <p>
        123{' '}
        <code>
          Hello, <b>R2D2!</b>
        </code>
        <br />
        You know what?
        <br />
        <i>
          <u>I'm your</u> <s>FATHER</s> <code>creator</code>.
        </i>
        <br />
        <pre>May the force be with you!</pre> Bye!
      </p>,
      null,
      null
    );

    await expect(promise).resolves.toMatchSnapshot();
    const segments = await promise;

    expect(segments?.map((r) => r.value)).toMatchInlineSnapshot(`
      [
        "123 Hello, R2D2!
      You know what?
      I'm your FATHER creator.
      May the force be with you! Bye!",
      ]
    `);
  });

  it('throw if non-textual node within children', async () => {
    const children = (
      <>
        foo
        <img />
        bar
      </>
    );

    await expect(
      renderer.render(<p>{children}</p>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <p/>"`
    );
    await expect(
      renderer.render(<b>{children}</b>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <b/>"`
    );
    await expect(
      renderer.render(<i>{children}</i>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <i/>"`
    );
    await expect(
      renderer.render(<s>{children}</s>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <s/>"`
    );
    await expect(
      renderer.render(<u>{children}</u>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <u/>"`
    );
    await expect(
      renderer.render(<code>{children}</code>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <code/>"`
    );
    await expect(
      renderer.render(<pre>{children}</pre>, null, null)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <img /> is placed in <pre/>"`
    );
  });

  test('should return null if content is empty', async () => {
    const elements = [<b />, <i />, <s />, <u />, <p />, <code />, <pre />];

    for (const element of elements) {
      await expect(renderer.render(element, null, null)).resolves.toEqual(null);
    }
  });
});

describe('media components', () => {
  it('metch snapshot', async () => {
    const medias = [
      <img src="http://..." />,
      <video src="http://..." />,
      <audio src="http://..." />,
      <file src="http://..." />,
    ];

    for (const media of medias) {
      await expect(renderer.render(media, null, null)).resolves.toEqual([
        {
          type: 'unit',
          node: media,
          value: {
            text: 'http://...',
            type: 'text',
          },
          path: '$',
        },
      ]);
    }
  });

  it('return "" if content is empty', async () => {
    const medias = [<img />, <video />, <audio />, <file />];

    for (const media of medias) {
      await expect(renderer.render(media, null, null)).resolves.toEqual([
        {
          type: 'unit',
          node: media,
          value: { text: '', type: 'text' },
          path: '$',
        },
      ]);
    }
  });
});
