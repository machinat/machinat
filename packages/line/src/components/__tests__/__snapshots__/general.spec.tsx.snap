// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<p/> <p/> renders into individual text segment 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$::0",
    "type": "text",
    "value": "foo",
  },
  Object {
    "node": "bar",
    "path": "$::1",
    "type": "text",
    "value": "bar",
  },
  Object {
    "node": <p>
      baz
    </p>,
    "path": "$::2",
    "type": "text",
    "value": "baz",
  },
]
`;

exports[`text components render nested elements 1`] = `
Array [
  Object {
    "node": <p>
      123
       
      <code>
        Hello, 
        <b>
          R2D2!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          creator
        </code>
        .
      </i>
      <br />
      <pre>
        May the force be with you!
      </pre>
       Bye!
    </p>,
    "path": "$",
    "type": "text",
    "value": "123 Hello, R2D2!
You know what?
I'm your FATHER creator.
May the force be with you! Bye!",
  },
]
`;

exports[`text components render shallow elements 1`] = `
Array [
  Array [
    Object {
      "node": <b>
        important
      </b>,
      "path": "$",
      "type": "text",
      "value": "important",
    },
  ],
  Array [
    Object {
      "node": <i>
        italic
      </i>,
      "path": "$",
      "type": "text",
      "value": "italic",
    },
  ],
  Array [
    Object {
      "node": <s>
        nooooo
      </s>,
      "path": "$",
      "type": "text",
      "value": "nooooo",
    },
  ],
  Array [
    Object {
      "node": <u>
        underlined
      </u>,
      "path": "$",
      "type": "text",
      "value": "underlined",
    },
  ],
  Array [
    Object {
      "node": <code>
        foo.bar()
      </code>,
      "path": "$",
      "type": "text",
      "value": "foo.bar()",
    },
  ],
  Array [
    Object {
      "node": <pre>
        foo.bar('hello world')
      </pre>,
      "path": "$",
      "type": "text",
      "value": "foo.bar('hello world')",
    },
  ],
  Array [
    Object {
      "node": <br />,
      "path": "$",
      "type": "text",
      "value": "
",
    },
  ],
]
`;
